website/docs/cdktf/examples-and-guides/refactoring.mdx

```diff
@@ -0,0 +1,255 @@
+
+## Moving resources inside a stack
+
+Let's say you have a stack deployed with an aws `S3Bucket` resource and you want to give it another name:
+
+```typescript
+import { Construct } from "constructs";
+import { App, TerraformStack } from "cdktf";
+import { AwsProvider } from "@cdktf/provider-aws/lib/provider";
+import { S3Bucket } from "@cdktf/provider-aws/lib/s3-bucket";
+
+class MyStack extends TerraformStack {
+  constructor(scope: Construct, id: string) {
+    super(scope, id);
+
+    new AwsProvider(this, "aws", {
+      region: "us-east-1",
+    });
+
+    // Old S3Bucket definition
+    // new S3Bucket(this, "my-bucket", {
+    //   bucketPrefix: "my-bucket",
+    // });
+
+    // New S3Bucket definition
+    new S3Bucket(this, "new-bucket", {
+      bucketPrefix: "my-bucket",
+    });
+  }
+}
+
+const app = new App();
+new MyStack(app, "refactoring-example");
+app.synth();
+```
+
+When you run a `cdktf plan` you might see the following output:
+
+```shell-session
+refactoring-example    # aws_s3_bucket.my-bucket will be destroyed
+                       # (because aws_s3_bucket.my-bucket is not in configuration)
+                       - resource "aws_s3_bucket" "my-bucket" {
+...
+refactoring-example    # aws_s3_bucket.new-bucket (new-bucket) will be created
+                       + resource "aws_s3_bucket" "new-bucket" {
+```
+
+As you can see the old bucket is going to be destroyed and a new one is created to replace it. This can be harmful if there is already state in the resource (e.g. files in the bucket, data in the database, etc).
+
+To avoid the recreation of the resource you can use the [`terraform state mv`](/cli/commands/state/mv) command to move the state before the next apply.
+
+```shell-session
+# before the change, find the terraform name of the resource
+# the name can be different than the name you put in the constructor of your construct,
+# see the information below
+$ cat cdktf.out/stacks/refactoring-example/cdk.tf.json | jq '.resource.aws_s3_bucket'
+{
+  "my-bucket": {
+    "//": {
+      "metadata": {
+        "path": "refactoring-example/my-bucket",
+        "uniqueId": "my-bucket"
+      }
+    },
+    "bucket_prefix": "my-bucket"
+  }
+}
+# the id is the key of the object, in this case it is "new-bucket"
+# if you have multiple resources of the same type, you can use the path to find the right one
+
+# Now change your code and synthesize it to get the new name of the resource
+$ cdktf synth # we need to know the new id of the resource
+$ cat cdktf.out/stacks/refactoring-example/cdk.tf.json | jq '.resource.aws_s3_bucket'
+{
+  "my-bucket": {
+    "//": {
+      "metadata": {
+        "path": "refactoring-example/new-bucket",
+        "uniqueId": "new-bucket"
+      }
+    },
+    "bucket_prefix": "new-bucket"
+  }
+}
+
+# Now that we know the old name is "my-bucket" and the new name is "new-bucket" we can move the state
+$ cd cdktf.out/stacks/refactoring-example
```

Comment:
This part should be highlighted that we're running the rest of the steps in the generated stack folder